pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go

```diff
@@ -258,6 +259,8 @@ func validateInterfaceNetworkBasics(field *k8sfield.Path, networkExists bool, id
 		causes = appendStatusCauseForSlirpNotEnabled(field, causes, idx)
 	} else if iface.Masquerade != nil && networkData.Pod == nil {
 		causes = appendStatusCauseForMasqueradeWithourPodNetwork(field, causes, idx)
+	} else if iface.Masquerade != nil && network.IsReserved(iface.MacAddress) {
```

Comment:
What will happen in case we have 1 interface with masq and the 2nd with bridge binding?
Initially I thought it may be a problem, but on the other hand it is on a different LAN. Will this cause problems with the ARP resolution? (I think we need to check this)

This also raised another point which I do not know how we can mitigate: If the secondary network is for example bridge-binding, this mac may just exists in the network (e.g. as the default-gateway), and in such a case we may also have a problem with the arp table (not sure if the ARP arp table accepts having a single mac mapped to two different IP addresses).

Can you check this?

Reply:
> What will happen in case we have 1 interface with masq and the 2nd with bridge binding?
> Initially I thought it may be a problem, but on the other hand it is on a different LAN. Will this cause problems with the ARP resolution? (I think we need to check this)

I don't see how, since those are different L2 networks.

> 
> This also raised another point which I do not know how we can mitigate: If the secondary network is for example bridge-binding, this mac may just exists in the network (e.g. as the default-gateway), and in such a case we may also have a problem with the arp table (not sure if the ARP arp table accepts having a single mac mapped to two different IP addresses).

Again, I don't see how. And furthermore, I don't see how this differs from a standard problem of having the same mac address in different L2 networks, which, I don't think is a problem.

> 
> Can you check this?

Never the less, I can. Below you can find the spec I used; it uses 2 VMs, both with the masquerade pod interface and a secondary macvtap network.

```yaml
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: macvtap
  annotations:
    k8s.v1.cni.cncf.io/resourceName: macvtap.network.kubevirt.io/eth0
spec:
  config: '{
      "cniVersion": "0.3.1",
      "type": "macvtap",
      "mtu": 1500
    }'
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: vm1
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: rootfs
      - disk:
          bus: virtio
        name: cloudinit
      interfaces:
      - masquerade: {}
        name: default
        ports:
        - name: ssh
          port: 22
      - macvtap: {}
        name: macvtap
        macAddress: 02:00:00:00:00:00
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  networks:
  - name: default
    pod: {}
  - multus:
      networkName: macvtap
    name: macvtap
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
    name: rootfs
  - cloudInitNoCloud:
      userData: |
        #!/bin/bash
        echo "fedora" |passwd fedora --stdin
        dhclient eth1
    name: cloudinit
---
apiVersion: kubevirt.io/v1alpha3
kind: VirtualMachineInstance
metadata:
  name: vm2
spec:
  domain:
    devices:
      disks:
      - disk:
          bus: virtio
        name: rootfs
      - disk:
          bus: virtio
        name: cloudinit
      interfaces:
      - masquerade: {}
        name: default
        ports:
        - name: ssh
          port: 22
      - macvtap: {}
        name: macvtap
    machine:
      type: ""
    resources:
      requests:
        memory: 1024M
  networks:
  - name: default
    pod: {}
  - multus:
      networkName: macvtap
    name: macvtap
  terminationGracePeriodSeconds: 0
  volumes:
  - containerDisk:
      image: registry:5000/kubevirt/fedora-cloud-container-disk-demo:devel
    name: rootfs
  - cloudInitNoCloud:
      userData: |
        #!/bin/bash
        echo "fedora" |passwd fedora --stdin
        dhclient eth1
    name: cloudinit
```

You can find below the macs / ips for the 2 VMs, along with their neighbor tables.

## VM1
```bash
$ [fedora@vm1 ~]$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:19:fa:ee brd ff:ff:ff:ff:ff:ff
    altname enp1s0
    inet 10.0.2.2/24 brd 10.0.2.255 scope global dynamic noprefixroute eth0
       valid_lft 86296654sec preferred_lft 86296654sec
    inet6 fe80::5054:ff:fe19:faee/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 02:00:00:00:00:00 brd ff:ff:ff:ff:ff:ff   <===== same MAC the gateway of eth0 has
    altname enp2s0
    inet 192.168.66.139/24 brd 192.168.66.255 scope global noprefixroute eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::e03c:bcdd:7d90:7e45/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever

# ping gateway - has 02:00:00:... mac address
[fedora@vm1 ~]$ ping 10.0.2.1
PING 10.0.2.1 (10.0.2.1) 56(84) bytes of data.
64 bytes from 10.0.2.1: icmp_seq=1 ttl=64 time=1.56 ms
64 bytes from 10.0.2.1: icmp_seq=2 ttl=64 time=0.927 ms

--- 10.0.2.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1003ms
rtt min/avg/max/mdev = 0.927/1.244/1.562/0.317 ms
[fedora@vm1 ~]$ ip n
192.168.66.131 dev eth1 lladdr 62:89:2e:cd:75:d7 STALE
10.0.2.1 dev eth0 lladdr 02:00:00:00:00:00 REACHABLE <===== as you can see here
192.168.66.2 dev eth1 lladdr 16:c1:6d:e8:62:dc STALE
fe80::b440:bff:fe6e:c1df dev eth1 lladdr 16:c1:6d:e8:62:dc router STALE
```

## VM2
```bash
[fedora@vm2 ~]$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:3f:06:10 brd ff:ff:ff:ff:ff:ff
    altname enp1s0
    inet 10.0.2.2/24 brd 10.0.2.255 scope global dynamic noprefixroute eth0
       valid_lft 86296796sec preferred_lft 86296796sec
    inet6 fe80::5054:ff:fe3f:610/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 62:89:2e:cd:75:d7 brd ff:ff:ff:ff:ff:ff
    altname enp2s0
    inet 192.168.66.131/24 brd 192.168.66.255 scope global noprefixroute eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::dcf9:be0:d53b:d725/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever

# 2 shells, one pinging vm1's IP - 192.168.66.139 - another pinging the gateway - 10.0.2.1
[fedora@vm2 ~]$ ping -q 192.168.66.139
[fedora@vm2 ~]$ ping -q 10.0.2.1

# gets us this neighbor table
[fedora@vm2 ~]$ ip n
192.168.66.2 dev eth1 lladdr 16:c1:6d:e8:62:dc STALE
192.168.66.139 dev eth1 lladdr 02:00:00:00:00:00 REACHABLE
10.0.2.1 dev eth0 lladdr 02:00:00:00:00:00 REACHABLE
fe80::b440:bff:fe6e:c1df dev eth1 lladdr 16:c1:6d:e8:62:dc router STALE
```

The ticks to update the neighbor state happen every `BASE_REACHABLE_TIME`/ 2, and the entries remain in the reachable state for a while, between `BASE_REACHABLE_TIME`/ 2 and 3/2 * `BASE_REACHABLE_TIME`  as seen in [the implementation](https://elixir.bootlin.com/linux/v5.6.6/source/net/core/neighbour.c#L958).

In the system above, w/ a 5.6.6. kernel, the `base_reachable_time` for the interfaces is 30 seconds:
```bash
[fedora@vm2 ~]$ cat /proc/sys/net/ipv4/neigh/eth0/base_reachable_time
30
[fedora@vm2 ~]$ cat /proc/sys/net/ipv4/neigh/eth1/base_reachable_time
30
```
I see that the neighbors persist in reachable state for around 10-30 seconds, which, given the numbers described above make sense to me.

And the traffic works, which I think is the most important metric here.

Reply:
Thanks!